pennylane_lightning/src/util/Macros.hpp

```diff
@@ -13,12 +13,155 @@
 // limitations under the License.
 /**
  * @file
- * Define some builtin alternatives
+ * Define macros and compile-time constants.
  */
 #pragma once
 
+#include <string>
+
+/**
+ * @brief Predefined macro variable to a string. Use std::format instead in
+ * C++20.
+ */
+#define PL_TO_STR_INDIR(x) #x
+#define PL_TO_STR(VAR) PL_TO_STR_INDIR(VAR)
+
 #if defined(__GNUC__) || defined(__clang__)
 #define PL_UNREACHABLE __builtin_unreachable()
-#else
+#elif defined(_MSC_VER)
 #define PL_UNREACHABLE __assume(false)
+#else // Unsupported compiler
+#define PL_UNREACHABLE
+#endif
+
+#if defined(__AVX2__)
+#define PL_USE_AVX2 1
+[[maybe_unused]] static constexpr bool use_avx2 = true;
+#else
+[[maybe_unused]] static constexpr bool use_avx2 = false;
+#endif
+
+#if defined(__AVX512F__)
+#define PL_USE_AVX512F 1
+[[maybe_unused]] static constexpr bool use_avx512f = true;
+#else
+[[maybe_unused]] static constexpr bool use_avx512f = false;
+#endif
+
+#if defined(__AVX512DQ__)
+#define PL_USE_AVX512DQ 1
+[[maybe_unused]] static constexpr bool use_avx512dq = true;
+#else
+[[maybe_unused]] static constexpr bool use_avx512dq = false;
+#endif
+
+#if defined(__AVX512VL__)
+#define PL_USE_AVX512VL 1
+[[maybe_unused]] static constexpr bool use_avx512vl = true;
+#else
+[[maybe_unused]] static constexpr bool use_avx512vl = false;
+#endif
+
+#if defined(_OPENMP)
+#define PL_USE_OMP 1
+[[maybe_unused]] static constexpr bool use_openmp = true;
+#else
+[[maybe_unused]] static constexpr bool use_openmp = false;
+#endif
+
+#if (_OPENMP >= 202011)
+#define PL_UNROLL_LOOP __Pragma("omp unroll(8)")
+#elif defined(__GNUC__)
+#define PL_UNROLL_LOOP _Pragma("GCC unroll 8")
+#elif defined(__clang__)
+#define PL_UNROLL_LOOP _Pragma("unroll(8)")
```

Comment:
Are `__Pragma` and `_Pragma` both valid in each case?

Reply:
Oops it must be `_Pragma`. I will fix it (I think it worked as OpenMP version `202011` is not really supported by any compilers).